core/vm.cpp

```diff
@@ -1291,6 +1293,51 @@ class Interpreter {
         return nullptr;
     }
 
+    const AST *builtinEncodeUTF8(const LocationRange &loc, const std::vector<Value> &args)
+    {
+        validateBuiltinArgs(loc, "encodeUTF8", args, {Value::STRING});
+
+        std::string byteString = encode_utf8(static_cast<HeapString *>(args[0].v.h)->value);
+
+        scratch = makeArray({});
+        auto &elements = static_cast<HeapArray *>(scratch.v.h)->elements;
+        for (const auto c : byteString) {
+            auto *th = makeHeap<HeapThunk>(idArrayElement, nullptr, 0, nullptr);
+            elements.push_back(th);
+            th->fill(makeNumber(uint8_t(c)));
+        }
+        return nullptr;
+    }
+
+    const AST *builtinDecodeUTF8(const LocationRange &loc, const std::vector<Value> &args)
+    {
+        validateBuiltinArgs(loc, "decodeUTF8", args, {Value::ARRAY});
+
+        std::string byteString;
+
+        HeapArray *arr = static_cast<HeapArray *>(scratch.v.h);
+        for (auto *thunk : arr->elements) {
+            if (!thunk->filled) {
+                stack.newCall(loc, thunk, thunk->self, thunk->offset, thunk->upValues);
```

Comment:
@sparkprime I need your help here. It doesn't work this way, and I'm not sure what is the right way. I need to get the (evaluated) elements of this array.

Reply:
I think you're going to need to record intermediate state on the stack while you recurse into the thunk elements. This is not easy.  Look at how std.filter is implemented natively -- it has a special stack frame it pushes.  If this puts you off implementing this natively in C++ I would understand it :)

Reply:
dmuir just did something similar for std.join in another PR